sys/usb/usbus/cdc/ecm/cdc_ecm_netdev.c

```diff
@@ -0,0 +1,198 @@
+    size_t len = iolist_size(iolist);
+    DEBUG("CDC_ECM_netdev: sending %u bytes\n", len);
+    /* load packet data into FIFO */
+    size_t iol_offset = 0;
+    size_t usb_offset = 0;
+    size_t usb_remain = cdcecm->ep_in->ep->len;
+    DEBUG("CDC_ECM_netdev: cur iol: %d\n", iolist->iol_len);
+    while (len) {
+        mutex_lock(&cdcecm->out_lock);
+        if (iolist->iol_len - iol_offset > usb_remain) {
+            /* Only part of the iolist can be copied, usb_remain bytes */
+            memcpy(buf + usb_offset, (uint8_t *)iolist->iol_base + iol_offset,
+                   usb_remain);
+
+            usb_offset =  cdcecm->ep_in->maxpacketsize;
+            len -= usb_remain;
+            iol_offset += usb_remain;
+            usb_remain = 0;
+        }
+        else {
+            size_t bytes_copied = iolist->iol_len - iol_offset;
+            /* Full iolist can be copied */
+            memcpy(buf + usb_offset, (uint8_t *)iolist->iol_base + iol_offset,
+                   bytes_copied);
+            len -= bytes_copied;
+            usb_offset += bytes_copied;
+            usb_remain -= bytes_copied;
+            iol_offset = iolist->iol_len;
+        }
+        if (iol_offset == iolist->iol_len) {
+            /* Current iolist exhausted */
+            iolist = iolist->iol_next;
+            if (iolist) {
+                DEBUG("CDC_ECM: cur iol: %d\n", iolist->iol_len);
+            }
+            iol_offset = 0;
+        }
+        if (usb_remain == 0 || !len) {
+            cdcecm->tx_len = usb_offset;
+            /* USB frame full or last frame, flush! */
+
+            DEBUG("CDC_ECM_NETDEV: triggering xmit with len %d\n",
+                  cdcecm->tx_len);
+            _signal_tx_xmit(cdcecm);
+            usb_remain = cdcecm->ep_in->maxpacketsize;
+            usb_offset = 0;
+        }
+        else {
+            mutex_unlock(&cdcecm->out_lock);
+        }
+    }
+    /* Zero length USB packet required */
+    if ((iolist_size(iolist_start) % cdcecm->ep_in->maxpacketsize) == 0) {
+        mutex_lock(&cdcecm->out_lock);
+        DEBUG("CDC ECM netdev: Zero length USB packet required\n");
+        cdcecm->tx_len = 0;
+        _signal_tx_xmit(cdcecm);
+    }
+    return len;
+}
+
+static int _recv(netdev_t *netdev, void *buf, size_t max_len, void *info)
+{
+    usbus_cdcecm_device_t *cdcecm = _netdev_to_cdcecm(netdev);
+
+    (void)info;
+    if (max_len == 0 && buf == NULL) {
+        return cdcecm->len;
+    }
+    if (max_len && buf == NULL) {
+        _signal_rx_flush(cdcecm);
+        return cdcecm->len;
+    }
+    memcpy(buf, cdcecm->in_buf, max_len);
+    _signal_rx_flush(cdcecm);
+    return max_len;
+}
+
+static int _init(netdev_t *netdev)
+{
+    usbus_cdcecm_device_t *cdcecm = _netdev_to_cdcecm(netdev);
+
+    luid_get(cdcecm->mac_netdev, ETHERNET_ADDR_LEN);
+    cdcecm->mac_netdev[0] |= 0x02;      /* locally administered address */
+    cdcecm->mac_netdev[0] &= ~0x01;     /* unicast address */
```

Comment:
blackmagic again ?

Reply:
Same as above, but now for the ethernet address on the host side

To explain, with ECM, we simulate a point to point link, on one side the host computer, on the other side the RIOT network interface. Both sides of the virtual network cable require a link layer address and with ECM the device specifies the host side link layer address via a descriptor. 

This bit generates and clamps the generated link layer address for the host side to indicate a locally administrated unicast link layer address.

Reply:
Changed to use the functions added in #11666